src/utils/csv.ts

```diff
@@ -0,0 +1,57 @@
+import { logDebug } from 'utils/miscellaneous'
+
+// TODO: can I make sure - using types - that headers length matches transformer return type length?
+// Alternatively, I could make headers be the keys of a second type parameter
+// which is the result of transformer, such as transformer<T, R>(item: T) => R[]
+interface ToCsvParams<T> {
+  headers?: string[]
+  data: T[]
+  transformer: (item: T) => string[]
+}
+
+/**
+ * Util function that transforms given data and transformer function into a csv string,
+ * taking care of escaping and error handling
+ *
+ * `headers` (optional). If present, length must match transformer return type length
+ * `data` is a list of items to be converted to CSV by `transformer`
+ * `transformer` is a function that receives the item and returns a list of strings
+ */
+export function toCsv<T>({ headers = [], data, transformer }: ToCsvParams<T>): string {
```

Comment:
I know you love to use fancy stuff, is this a job for generators? 

You don't need to change this, only if it really makes it more robust  (if it makes a difference), i.e. for big files, instead of creating all the array of string and joining you process the elements and create the blob from the steam.

Reply:
I don't understand what you are suggesting.
How would you stream to a blob, I don't think you can
Neither can you stream from a blob.
And `URL.createObjectURL` is sync